Test that `slotStartTime . slotRange` is idempotent.

Anviking opened the pull request:
# Issue Number

#466 

# Overview

This PR:

- [x] adds a property to check that the composition of `slotStartTime` with `slotRange` is idempotent.
- [x] updates the documentation for `Range` to use idiomatic range boundaries for negative and positive infinity.

Anviking committed:
Change \infty bounds to be exclusive

Anviking committed:
Add slotStartTime and slotRange properties

Anviking commented on lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -659,6 +667,11 @@ makeRangeValid = \case
 
+rangeIsSubrangeOf :: Ord a => Range a -> Range a -> Bool
+rangeIsSubrangeOf (Range a b) r =
+    maybe True (`isWithinRange` r) a
```
The `True` case is wrong though I think 🙃

jonathanknowles replied:
I think you're right to be suspicious!

For example, the following expression evaluates to `True`:

```hs
% Range Nothing Nothing `rangeIsSubrangeOf` Range (Just 0) (Just 0)
True
```

I've made a PR here with an implementation of `isSubrangeOf` (and property tests):

https://github.com/input-output-hk/cardano-wallet/pull/653

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -369,11 +369,11 @@ instance FromText SortOrder where
 -- +---------------------------------+-------------+---------------------------+
--- | @'Range' 'Nothing'  'Nothing' @ | @[−∞, ∞ ]@  | @\\p -> True            @ |
+-- | @'Range' 'Nothing'  'Nothing' @ | @(−∞, ∞ )@  | @\\p -> True            @ |
 -- +---------------------------------+-------------+---------------------------+
```
:+1:

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -362,6 +362,14 @@ spec = do
+                    let r = getUniformTime <$> timeRange
+                    case fmap (slotStartTime sps) <$> slotRange sps r of
+                        Just r' -> r' `rangeIsSubrangeOf` r
+                        Nothing -> True
```
I think this is a great idea for a property.

Though perhaps we can make it even stronger.

Can you think of a way to test that `slotRange` gives the **_maximal_** result?

I can think of two properties to test:

* the lower slot bound (if present) should not be higher than necessary.
* the upper slot bound (if present) should not be lower than necessary.

Anviking replied:
I like where you're going!

I think a naive approach would lead to re-implementing `slotRange`, which would be undesirable.

There is a property very similar to (imagined) user-requirements for pagination I had imagined would be satisfied, but it is not!

If I have `r1 = [a, b]`, `r2 = [b,c]` I'd imagined that the union of calling `listTransactions` with `r1` and `r2` would be the same as calling `listTransactions` with `[a,c]` once.

I named `f = fmap slotStartTime . slotRange`, so more formally I'd expect something like `union $ map f (split timeRange n) === f timeRange`

This is actually an unlikely case since `f` shrinks ranges. We might get gaps, or we might even only get a few singleton ranges! 😱

<img width="914" alt="Skärmavbild 2019-08-22 kl  13 06 28" src="https://user-images.githubusercontent.com/304423/63513284-afeef380-c4dd-11e9-90ee-b3d3301e1c9d.png">

Anviking replied:
So things went wrong in #595 already 😐 🔫

Anviking replied:
**No actually, this might be as intended since transactions only have times inside the singleton ranges**

Anviking replied:
Tried two similar methods for a property, very messy still.

way to go might might be to generate `sps, slots, ratio`, let `[t1, t3]` be the range that encloses all slots and place `t2` in between based on the `ratio`.

jonathanknowles reviewed (changes requested):
I think this a great idea for a test. Though I suspect we can make it even stronger by checking that the range produced is _**maximal**_ (see review comment for more details).

As for the `isSubrangeOf` function. I've created a PR to implement this here (along with tests): #653. Would be interested to hear your feedback on that.

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -362,6 +362,14 @@ spec = do
                             . slotStartTime sps
                     Just slot === f slot
 
+        it "slotStartTime (slotRange timeRange) ⊆ timeRange " $
```
Small point, but perhaps this should be `"fmap slotStartTime (slotRange timeRange) ⊆ timeRange"`

 What do you think?

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -362,6 +362,14 @@ spec = do
+                \sps timeRange -> do
+                    let r = getUniformTime <$> timeRange
+                    case fmap (slotStartTime sps) <$> slotRange sps r of
+                        Just r' -> r' `rangeIsSubrangeOf` r
```
Do you think we should add some coverage analysis here to ensure that we cover both the `Just` and the `Nothing` cases?

We can use `cover` and `checkCoverage` to do this. (See `"arbitrary ranges are valid"` for an example.)

Anviking replied:
:+1: with some restructuring it was not that awkward to do

I don't think we care about covering the `Nothing` case though.

jonathanknowles replied:
> +1 with some restructuring it was not that awkward to do
> 
> I don't think we care about covering the `Nothing` case though.

Function `slotRange` should only produce `Nothing` when the upper bound is earlier than the start of the blockchain. So if we get a `Nothing`, we should check that this condition is satisfied.

Anviking replied:
In my eyes the `"f timeRange ⊆ timeRange"` doesn't care about _why_ or _when_ `slotRange` returns `Nothing`, which is tested elsewhere

jonathanknowles replied:
Agree that this is tested elsewhere. :+1:

jonathanknowles commented on lib/test-utils/src/Test/Utils/Time.hs:
```diff
@@ -27,6 +27,9 @@ newtype UniformTime = UniformTime { getUniformTime :: UTCTime }
     arbitrary = UniformTime <$> genUniformTime
+    shrink x = if x == newyear2000 then [] else [newyear2000]
+      where
+        newyear2000 = UniformTime $ UTCTime (ModifiedJulianDay 51544) 0
```
Why is the year 2000 special here?

jonathanknowles replied:
Is there any value in shrinking values of `UniformTime`?

Anviking replied:
At the moment when you reviewed: I'm not sure.

`StartTime` interacts with the `SlotId` to `UTCTime` conversions.

When debugging test-failures I got tired of having to parse extreme-dates and figure out if they were before or after other extreme dates, and if the extremeness of those dates mattered. So I wanted e.g. the `StartTime` to shrink to some easily identifiable value, if it had nothing to do with the failure. 

At that point in time  it seemed to help a bit, but not as reliably as I'd liked 🤷‍♂

Anviking replied:
I'll revert this while reverting the third property.

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -447,6 +453,51 @@ spec = do
+                \sps timeRange (NonNegative n) -> do
+                    let r = getUniformTime <$> timeRange
+                    let r' = f sps r
+                    iterate (>>= f sps) r' !! n === r'
```
Does the existence of `Data.Function.Utils.applyN` make the above property easier to define?

Anviking replied:
Reads better at least 👍

jonathanknowles replied:
Great!

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -447,6 +453,51 @@ spec = do
+                    iterate (>>= f sps) r' !! n === r'
+
+        it "slots in `slotRange [t1, t3]` are also in one or more of \
+           \`slotRange [t1, t2]` or `slotRange [t2, t3]` if t1 < t2 < t3" $
```
Regarding the following property:

> "slots in `slotRange [t1, t3]` are also in one or more of `slotRange [t1, t2]` or `slotRange [t2, t3]` if t1 < t2 < t3"

If `t2` corresponds to the start time of slot `s`, and if slot `s` has transactions, then those transactions will indeed occur in both `slotRange t1 t2` and `slotRange t2 t3` (as our range filtering is inclusive).

jonathanknowles reviewed (changes requested):
I think that defining extra properties to detect overlaps between ranges will be very valuable when we eventually build support paging.

However, given that we don't have this requirement just yet, should we define properties to that end?

One thing I think is useful from this PR is to assert that: `f r == f (f r)`, as you have done. In other words, that the `fmap startTime . slotRangeFromTimeRange` function is idempotent, and that repeated applications don't produce a smaller range.

Anviking committed:
Remove unfinished property about splitting slots into multiple time-ranges

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -506,6 +508,23 @@ spec = do
+        let f sps r = fmap (slotStartTime sps) <$> slotRangeFromTimeRange sps r
+        let sharedDescription = " (where f = fmap slotStartTime . \
+            \slotRangeFromTimeRange)"
+        it ("f timeRange ⊆ timeRange" ++ sharedDescription) $
```
Hi @Anviking. This is a good test, and we should definitely have it. Though I'm fairly certain it duplicates "Rule 1" in the test above:

https://github.com/input-output-hk/cardano-wallet/pull/652/files#diff-ea01d42dc8106d3d2e18f8ec129c9fa3R494-R496

What do you think?

jonathanknowles replied:
I've checked, and I'm pretty sure that this property duplicates rule 1 from the existing property. So I've removed this property for now.

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -506,6 +508,23 @@ spec = do
+                    case f sps r of
+                        Just r' -> r' `isSubrangeOf` r
+                        Nothing -> True
+        it ("f^n === f" ++ sharedDescription) $
```
What do you think about the following name:

```hs
it "(fmap (fmap slotStartTime) . slotRangeFromTimeRange) is idempotent"
```
One advantage is that the user doesn't have to scroll up to determine what `f` is, as the description is completely self-contained.

Another advantage is that "idempotent" already has a well-defined meaning here.

What do you think?

(I've omitted the `SlotParameter` arguments for conciseness.)

jonathanknowles replied:
I've changed this to "`slotStartTime . slotRangeFromTimeRange` is idempotent". We're using `.` in a very hand-wavy way, but testing the composition of these two functions is indeed our intent.

jonathanknowles committed:
Remove property that duplicates part of existing property.

jonathanknowles committed:
Inline value of `sharedDescription` and tweak wording.

jonathanknowles committed:
Move definition of `f` to within the property that uses it.

jonathanknowles committed:
Inline definition of `r'`.

jonathanknowles committed:
Remove unnecessary imports and references to unused language extensions.

jonathanknowles committed:
Remove use of `applyN`.

It's not necessary to apply a pure function `f` multiple times to check that
for idempotency. It's enough to check that applying it twice gives the
same result as applying it once.

jonathanknowles committed:
Add coverage check to property.

rvl reviewed (approved):
OK!
